### PR TITLE
Asset updates

### PR DESCRIPTION
**Pull request by einarmo**
Implementation of retry logic for asset updates. This is probably the most difficult resource to handle updates for, since the API is inconsistent in some complicated ways, especially related to parent ids. This handles most cases, but not all, some of the weirder edge cases are probably missed.

### CODE REVIEW

File: Cognite.Extensions/Sanitation.cs [hunk 1]
```diff
@@ -54,6 +54,36 @@ private static bool CheckLength(this string? str, int maxLength)
...
         }
 
+        /// <summary>
+        /// Reduce the length of "set" on given update to maxLength, if it is longer.
+        /// </summary>
+        /// <param name="update">String update to be shortened</param>
+        /// <param name="maxLength">Maximum length of final string</param>
+        /// <returns>String update which contains the first <paramref name="maxLength"/> characters of the passed value.</returns>
+#if NETSTANDARD2_1_OR_GREATER
+        [return: System.Diagnostics.CodeAnalysis.NotNullIfNotNull("id")]
```
Comment (silvavelosa) [inline Cognite.Extensions/Sanitation.cs:64]: shouldn't this be "update"?

**Review by silvavelosa — APPROVED** [review]
Cool! just a couple questions

File: Cognite.Extensions/Assets/AssetExtensions.cs [hunk 2]
```diff
@@ -330,5 +330,131 @@ await Task
...
+                    results[idx] = result;
+                });
+
+            int taskNum = 0;
+            await generators.RunThrottled(
+                throttleSize,
+                (_) => {
+                    if (chunks.Count > 1)
+                        _logger.LogDebug("{MethodName} completed {NumDone}/{TotalNum} tasks",
+                            nameof(UpdateAsync), ++taskNum, chunks.Count);
```
[the remark this answers is not in the document]
Comment (einarmo) [reply Cognite.Extensions/Assets/AssetExtensions.cs:392]: I think you could, but it is unlikely and it only affects logging. I'm not convinced locking is worthwhile.

[the remark this answers is not in the document]
Comment (einarmo) [reply Cognite.Extensions/Assets/AssetExtensions.cs:392]: Ah, no you can't, the callback runs on a single thread.

**Pull request merged**